### DISCUSSION TITLE
rabbitmq - When does it page out to disk?

### DISCUSSION THREAD

**Discussion by luarx** [rabbitmq/rabbitmq-server#10992] [Other] [answered]
```
- RabbitMQ version: 3.13.1
- Docker image: rabbitmq:3
- Calculation strategy: rss
- Memory high watermark setting: 0.4 of available memory, computed to: 6.5492 gb
- How runs: Docker image in Kubernetes
- Node host architecture: ARM
```

First of all, **excuse me for opening this discussion as maybe it is trivial but I can not find in rabbitmq docs information about this** (if there is , just point me out to the right place)🙏  

Afaik, rabbitmq considers `vm_memory_high_watermark` and `vm_memory_high_watermark_paging_ratio` values (this lask one maybe [obsolete](https://www.rabbitmq.com/docs/memory#paging)? ) to know when it needs to page out to disk.

My use case is that from time to time:
- My rabbitmq container starts to page out to disk and it is not able to revert the situation unless I restart it and delete all memory/disk data. 
  - **When that happens, memory usage is lower than** `vm_memory_high_watermark` and `vm_memory_high_watermark_paging_ratio`
- Because of the pagination (or maybe it is in the other way round) rabbitmq is slower

I would appreciate if you could point me out in the right direction 🙏 🚀 

I share screenshots of a real situation that started at 9am:
- CPU/Memory resources were enough and lower than the `vm_memory_high_watermark` values (Memory usage: 300Mi  and vm_memory_high_watermark: 6.5492 gb) but rabbitmq started to page out to disk when the memory usage increased a little bit
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/6ed6e407-24be-4fb4-9aa9-5bd125495f3a)
- When the previous point happened, rabbitmq server paged out to disk:
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/b537854a-2a0d-4d24-a83a-c67ad6449dd5)
- And here you have some more metrics around that moment:
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/1614f870-c9de-443b-b180-083f0b197e2b)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/f94cd6d1-ae0f-4fbf-a151-2b8b0ab31efa)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/181ac6cb-0824-48d4-b2c0-92e08512490d)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/4fe71053-4c37-4507-8ccd-0088810e42d8)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/42ef987a-b5cf-46eb-8d1f-6acf47c93eae)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/98d61b2c-0f65-43d2-9965-a586e8b9e8db)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/10a8756e-754c-4da4-88cb-1a66978df451)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/7e6a5d45-0ab2-43f2-9f3d-2622a3f4ceb1)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/7d2412e1-97d4-47ac-9e0f-d8b0bdea3ca3)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/dca773a7-54fa-494b-975f-f4c59340b0c7)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/4c6db9c4-9db5-41df-b03d-b2af5995b85b)
![image](https://github.com/rabbitmq/rabbitmq-server/assets/22997139/b58970ba-0bad-47c9-82ce-59cf0b868282)

**Comment by michaelklishin** [comment]
`vm_memory_high_watermark_paging_ratio` is only relevant for classic queues v1, not classic queues v2 (CQv2).

Quorum queues move data to disk immediately but [keep a certain number of WAL entries](https://www.rabbitmq.com/docs/quorum-queues#performance-tuning) in memory.

Streams always move data to disk "straight from the socket" and keep very little in memory.

Classic queues v2 act as classic queues v1 in lazy mode if we oversimplify, that is, they generally move data to disk quickly.

**Comment by michaelklishin** [comment, marked as answer]
I don't know how to interpret these charts, "message pipeline" is not a metric that RabbitMQ exposes and not a part of our own Grafana dashboards. According to the chart, there is a massive backlog build-up that then very quickly clears, possibly because a consumer shows up.

There can be any number of reasons for that, and "slow RabbitMQ" very often turns out to be a slow group of consumers which publishers easily outpace. One way or another resource use (even if you use streams) will grow until there's enough consumers online or publishers are throttled heavily enough to make the inflow/outflow more balanced.

There are also [storage layer recommendations](https://www.rabbitmq.com/docs/production-checklist#storage) and [certain tuning recommendations](https://www.rabbitmq.com/docs/quorum-queues#performance-tuning) for quorum queues that may benefit the QQs running at close to their peak throughput.

We have no details on the queue types used, their durability, on whether applications publish messages as transient (the latter two are important for CQv1s) to offer an informed hypothesis. We do no guess in this community.

When in doubt, use quorum queues, streams and CQv2s, and fast disks.

**Reply by luarx** [reply]
Thank you for your answer! Will take a look and see if I can debug further with your advises 🙏
